### PR TITLE
Tests are now being executed during gcc builds

### PR DESCRIPTION
**Pull request by dmitri-d**
This [re]enables tests during gcc builds.

Please note this PR shouldn't ve merged before https://github.com/envoyproxy/envoy-build-tools/pull/34.

I'm not sure if it's ok to use ```bazel.gcc``` target, or a new target should be added, decided to use ```bazel.gcc``` in this PR.

### CODE REVIEW

**Comment by lizan** [comment]
@dmitri-d merged envoyproxy/envoy-build-tools#34

**Comment by dmitri-d** [comment]
@lizan: could you restart the tests please?

**Comment by dmitri-d** [comment]
There's a bunch of failures due to memory leaks:

```
*** WARNING: Cannot convert addresses to symbols in output below.
*** Reason: Cannot find 'pprof' (is PPROF_PATH set correctly?)
*** If you cannot fix this, try running pprof directly.
Leak of 8 bytes in 1 objects allocated from:
	@ 40ac9c 
	@ b4b8ad 
```
I tracked them down to static initialiser here: https://github.com/envoyproxy/envoy/blob/master/source/extensions/common/crypto/utility_impl.cc#L105. Not sure what to do about these other than to pass "HEAPCHECK=minimal" to tests.

**Comment by dmitri-d** [comment]
ping

**Comment by lizan** [comment]
@dmitri-d you need to do the build image update as well to pick up gcc-9, so this can't be in until #10506 is merged.

**Comment by dmitri-d** [comment]
@lizan: could you restart the tests pls?

**Comment by lizan** [comment]
@dmitri-d you need merge master.

**Comment by dmitri-d** [comment]
* merged in master

**Comment by lizan** [comment]
some QUIC test failed to build in gcc. cc @danzh2010

**Comment by dmitri-d** [comment]
@lizan: the build fails due to https://github.com/envoyproxy/envoy/issues/10422.

**Comment by dmitri-d** [comment]
@lizan: could you re-open the PR please? At least one of the quiche-related issues has been resolved.

**Comment by dmitri-d** [comment]
@lizan -- not sure what to do with these failures, any thoughts? There are a *lot* of memory leaks reported, and I think I have a few of these tests passing locally just fine.

**Comment by lizan** [comment]
@dmitri-d seems its the heap profiler detected some leak, not sure whether its a false positive. If you think it's false positive, add `--test_env=HEAPCHECK=`?

**Comment by dmitri-d** [comment]
@lizan: pprof points to https://github.com/envoyproxy/envoy/blob/master/test/fuzz/fuzz_runner.cc#L56 as the source of the leak. I can try disabling heapcheck for log_silencer, but it would require dragging in pprof as a dependency.

**Comment by lizan** [comment]
@dmitri-d don't even specify `minimal` I meant just do `--test_env=HEAPCHECK=` literally to disable it.

**Comment by dmitri-d** [comment]
MacOS tests are failing today, and one shard of //test/integration:protocol_integration_test failed, which I think is a bit flaky? WDYT, @lizan?

**Comment by dmitri-d** [comment]
* merged in latest changes

**Comment by dmitri-d** [comment]
* Hitting https://github.com/envoyproxy/envoy/issues/7647 in dynamic_opentracing_driver_impl_test
* In quic_platform_test, the stacktrace generated with `-c opt` doesn't include the test where the exception was thrown
* stats_integration_test reports different values under gcc
* MacOS tests continue to flake

@lizan -- I think you had a fix here (https://github.com/envoyproxy/envoy/compare/master...lizan:dynamic_ot?expand=1) for the dynamic_opentracing_driver_impl_test failure? Not sure what effect it will have on clang/libc++ toolchain though. As an alternative, I can disable this test under gcc altogether. 
Not sure what to do re: quic_platform_test, except disable it under gcc.
stats_integration_test: I can drop `ENVOY_MEMORY_TEST_EXACT=true` when running tests under gcc.

**Comment by lizan** [comment]
@dmitri-d

- feel free to reuse my patch (better upstream to opentracing though). I don't think it affect clang/libc++.
- for quick-platform-test, I'm fine with ifdef there too.
- for stats_integration_test, yes let's drop that. they will never match.

**Comment by dmitri-d** [comment]
* dynamic_opentracing_driver_impl_test is passing now
* disabled quic_platform_test
* under gcc stats_integration_test uses Mode::Approximate even when `ENVOY_MEMORY_TEST_EXACT=true`

**Comment by dmitri-d** [comment]
Tried using https://github.com/envoyproxy/envoy/compare/master...lizan:dynamic_ot?expand=1, but it breaks asan (libc++ and tcmalloc in general?) dynamic_ot:dynamic_opentracing_driver_impl_test and dynamic_ot:config_test tests with:

```
C++ exception with description "opentracing: failed to load dynamic library: /b/f/w/bazel-out/k8-dbg/bin/test/extensions/tracers/dynamic_ot/dynamic_opentracing_driver_impl_test.runfiles/io_opentracing_cpp/mocktracer/libmocktracer_plugin.so: cannot allocate memory in static TLS block" thrown in the test body.
```

As much as I dislike it, I think disabling these tests under gcc is the way forward.

@lizan

**Comment by lizan** [comment]
@dmitri-d sgtm

**Comment by dmitri-d** [comment]
@lizan: the build is green after disabling dynamic_ot:dynamic_opentracing_driver_impl_test

File: test/common/stats/stat_test_utility.cc [hunk 1]
```diff
@@ -119,20 +119,23 @@ MemoryTest::Mode MemoryTest::mode() {
   const size_t end_mem = Memory::Stats::totalCurrentlyAllocated();
   bool can_measure_memory = end_mem > start_mem;
 
+// memory usage is different under gcc, we use Mode::Approximate is possible
```
Comment (lizan) [inline test/common/stats/stat_test_utility.cc:122]: You don't need this change, just don't set the env in gcc CI?

File: ci/do_ci.sh [hunk 2]
```diff
@@ -137,9 +137,13 @@ elif [[ "$CI_TARGET" == "bazel.sizeopt" ]]; then
   bazel test ${BAZEL_BUILD_OPTIONS} --config=sizeopt ${TEST_TARGETS}
   exit 0
 elif [[ "$CI_TARGET" == "bazel.gcc" ]]; then
+  BAZEL_BUILD_OPTIONS="${BAZEL_BUILD_OPTIONS} --test_env=ENVOY_MEMORY_TEST_EXACT=true --test_env=HEAPCHECK="
```
Comment (lizan) [inline ci/do_ci.sh:140]: You don't need `--test_env=ENVOY_MEMORY_TEST_EXACT=true`

**Comment by dmitri-d** [comment]
* Removed ENVOY_MEMORY_TEST_EXACT from the gcc target
* merged in the latest master

**Comment by dmitri-d** [comment]
@lizan: could you restart the build please? There were networking issues during the last run.

**Comment by twghu** [comment]
Looking at errors from gcc tests.

**Comment by twghu** [comment]
Refer to https://github.com/envoyproxy/nighthawk/issues/382#issuecomment-650179207 for explanation of problem.

File: test/mocks/config/mocks.h [hunk 3]
```diff
@@ -94,6 +94,10 @@ class MockGrpcMux : public GrpcMux {
   MockGrpcMux();
   ~MockGrpcMux() override;
 
+  using GrpcMux::pause;
```
Comment (lizan) [inline test/mocks/config/mocks.h:97]: why did you need this? Does marking MOCK_METHOD below override in 4th parameter (like below) fixes gcc build?

```
MOCK_METHOD(void, start, (), (override));
```

File: source/common/common/macros.h [hunk 4]
```diff
@@ -54,4 +54,7 @@ namespace Envoy {
 #define FALLTHRU
 #endif
 
+#if (defined(__GNUC__) && !defined(__clang__))
```
Comment (lizan) [inline source/common/common/macros.h:57]: This should be in include/envoy/common/platform.h?

Comment (dmitri-d) [reply source/common/common/macros.h:57]: Hmm, the comment in platform.h says:
```
"platform.h" header exists to simplify the most common references
to non-ANSI C/C++ headers, required on Windows, Posix, Linux, BSD etc,
and to provide substitute definitions when absolutely required.
``` 
feels like not a good spot to me for the macro in question?

**Comment by dmitri-d** [comment]
* Added 'override' to MockGrpcMux methods implementing the interface

File: source/common/common/macros.h (hunk 4 above)
Comment (lizan) [reply source/common/common/macros.h:57]: OK, I don't feel strongly either.

**Comment by dmitri-d** [comment]
@lizan: Not sure why asan failed (all tests passed), otherwise looks good.

**Comment by dmitri-d** [comment]
* resolved conflict, merged in latest changes

**Comment by dmitri-d** [comment]
@lizan, the build is green now.

**Comment by dmitri-d** [comment]
Thank you very much for the reviews and feedback, @lizan.
